libs/error-stack/src/macros.rs

```diff
@@ -190,13 +190,96 @@ macro_rules! report {
 /// }
 /// # Ok(())
 /// ```
+#[cfg(not(feature = "unstable"))]
 #[macro_export]
 macro_rules! bail {
-    ($err:expr $(,)?) => {{
+    ($err:expr) => {{
         return $crate::Result::Err($crate::report!($err));
     }};
 }
 
+/// Creates a [`Report`] and returns it as [`Result`].
+///
+/// Shorthand for `return `Err`(`[`report!(...)`]`)`
+///
+/// [`Report`]: crate::Report
+/// [`report!(...)`]: report
+///
+/// # Examples
+///
+/// Create a [`Report`] from [`Error`]:
+///
+/// [`Error`]: core::error::Error
+///
+/// ```
+/// use std::fs;
+///
+/// use error_stack::bail;
+/// # fn wrapper() -> error_stack::Result<(), impl core::fmt::Debug> {
+/// match fs::read_to_string("/path/to/file") {
+///     Ok(content) => println!("file contents: {content}"),
+///     Err(err) => bail!(err),
+/// }
+/// # Ok(()) }
+/// # assert!(wrapper().unwrap_err().contains::<std::io::Error>());
+/// ```
+///
+/// Create a [`Report`] from [`Context`]:
+///
+/// [`Context`]: crate::Context
+///
+/// ```rust
+/// # fn has_permission(_: &u32, _: &u32) -> bool { true }
+/// # type User = u32;
+/// # let user = 0;
+/// # type Resource = u32;
+/// # let resource = 0;
+/// use core::fmt;
+///
+/// use error_stack::{bail, Context};
+///
+/// #[derive(Debug)]
+/// # #[allow(dead_code)]
+/// struct PermissionDenied(User, Resource);
+///
+/// impl fmt::Display for PermissionDenied {
+///     # #[allow(unused_variables)]
+///     fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+///         # const _: &str = stringify! {
+///         ...
+///         # }; Ok(())}
+/// }
+///
+/// impl Context for PermissionDenied {}
+///
+/// if !has_permission(&user, &resource) {
+///     bail!(PermissionDenied(user, resource));
+/// }
+/// # Ok(())
+/// ```
+#[cfg(feature = "unstable")]
```

Comment:
Is there a way to not include this feature flag in the docs? Maybe somethig like `#[cfg(not(docs))]`? Because currently, we have this:
<img width="183" alt="image" src="https://github.com/user-attachments/assets/bdd30ebf-86ed-492f-a0cb-355d8f4aab1f">

Reply:
I would've liked if there's a way to conditionally disable match arms that would circumvent the problem. Sadly I didn't find an obvious ergonomic way to do so. I'll just disable the unstable macro during generation.

Reply:
I noticed, that the `bail` macro is probably not working as expected:
```rust
error[E0034]: multiple applicable items in scope
  --> libs/error-stack/tests/test_macros.rs:36:55
   |
36 |     let report = capture_error(|| bail!([RootError])).attach_printable(PrintableA(0));
   |                                                       ^^^^^^^^^^^^^^^^ multiple `attach_printable` found
   |
   = note: candidate #1 is defined in an impl for the type `error_stack::Report<C>`
   = note: candidate #2 is defined in an impl for the type `error_stack::Report<[C]>`
```

Probably an issue for a follow-up (-> non-blocking)

Reply:
I tried around with it a bit and failed to properly generate a doc here. We might want to remove the current `bail` from the docs and add a `bail` macro with both variants and a notice that the second variant is `unstable`.

Reply:
I noticed this as well, I wanted to do this in a follow up - because the code for `Report<[C]>` and `Report<C>` is actually the same now (previously the return type was different)